File: constraints/dataAnalysis/stellarMassFunction_SDSS_z0.07/massLuminosityRelation.pl
#!/usr/bin/env perl
use strict;
use warnings;
my $galacticusPath;
if ( exists($ENV{"GALACTICUS_ROOT_V093"}) ) {
 $galacticusPath  = $ENV{"GALACTICUS_ROOT_V093"};
 $galacticusPath .= "/" unless ( $galacticusPath =~ m/\/$/ );
} else {
 $galacticusPath  = "./";
}
unshift(@INC,$galacticusPath."perl"); 
use PDL;
use PDL::NiceSlice;
use PDL::IO::Misc;
use PDL::Fit::Polynomial;
use Astro::Cosmology;
require Stats::Percentiles;
require GnuPlot::PrettyPlots;
require GnuPlot::LaTeX;

# Determine the relation between stellar mass and SDSS r-band absolute magnitude using semi-analytic models from the Millennium
# Database. Specifically, the DeLucia2006a models are used - these correspond to De Lucia & Blaizot (2007;
# http://adsabs.harvard.edu/abs/2007MNRAS.375....2D) which use a Chabrier IMF, consistent with the IMF used by Li & White (2009).
# Andrew Benson (10-July-2012)

# Define a working directory.
my $workDirectory = $galacticusPath."constraints/dataAnalysis/stellarMassFunction_SDSS_z0.07/massLuminosityWork/";
system("mkdir -p ".$workDirectory);

# Define little Hubble parameter for the Millennium Simulation.
my $hubble = pdl 0.7;

# Define the limiting apparent magnitude of the Li & White (2009) sample.
my $apparentMagnitudeLimit = pdl 17.6;

# Define Millennium Simulation cosmology.
my $cosmology = Astro::Cosmology->new(
    omega_matter => 0.3,
    omega_lambda => 0.7,
    H0           => 70.0
    );
 
# Define the database URL.
my $databaseURL = "http://gavo.mpa-garching.mpg.de/Millennium/?action=doQuery&SQL=";

# Get list of snapshot numbers and corresponding redshifts.
my $sqlQuery = "select snapnum, z from millimil..Snapshots";
system("wget \"".$databaseURL.$sqlQuery."\" -O ".$workDirectory."millenniumDB_snapshots.csv")
    unless ( -e $workDirectory."millenniumDB_snapshots.csv" );
(my $snapshotNumbers, my $redshifts) = rcols($workDirectory."millenniumDB_snapshots.csv",{EXCLUDE => '/(^#|snapnum)/', COLSEP => ","});

# Initialize vectors to store results.
my $redshiftTable     = pdl ( 0.0 );
my $limitingMassTable = pdl ( 6.0 );

# Loop over redshifts.
for(my $iSnapshot=nelem($snapshotNumbers)-1;$iSnapshot>=0;--$iSnapshot) {
    my $redshift = $redshifts->(($iSnapshot));

    # Only consider reasonably low redshifts.
    if ( $redshift <= 0.5 && $redshift > 0.0 ) {

	# Compute the limiting absolute magnitude at this redshift.
	my $limitingMagnitude = $cosmology->absolute_magnitude($apparentMagnitudeLimit,$redshift);

	# Define the SQL query.
	my $sqlQuery = "select mass.stellarMass, light.r_sdss from millimil..DeLucia2006a as mass, millimil..DeLucia2006a_sdss2mass as light where mass.snapnum = ".$iSnapshot." and mass.galaxyID = light.galaxyID";

	# Download the data.
	system("wget \"".$databaseURL.$sqlQuery."\" -O ".$workDirectory."millenniumDB_".$iSnapshot.".csv")
	    unless ( -e $workDirectory."millenniumDB_".$iSnapshot.".csv" );

	# Read the data.
	(my $mass, my $magnitude) = rcols($workDirectory."millenniumDB_".$iSnapshot.".csv",{EXCLUDE => '/(^#|stellarMass)/', COLSEP => ","});

	# Convert mass to logarithmic scale and Solar units.
	my $logMass = log10(1.0e10*$mass/$hubble);

	# Define weights.
	my $weight = pdl ones(nelem($logMass));

	# Define mass bins.
	my $logMassMinimum = pdl  8.0;
	my $logMassMaximum = pdl 13.0;
	my $logMassCount   = 10;
	my $logMassBins    = sequence($logMassCount)*($logMassMaximum-$logMassMinimum)/($logMassCount-1)+$logMassMinimum;

	# Find the median magnitude as a function of mass.
	my $percentiles = pdl ( 50.0 );
	my $median      = &Percentiles::BinnedPercentiles($logMassBins,$logMass,$magnitude,$weight,$percentiles);

	# Fit a polynomial to the results.
	my $nonZeroBins = which($median->(:,(0)) < 0.0);
	(my $fit, my $coeffs) = fitpoly1d($median->(:,(0))->index($nonZeroBins),$logMassBins->index($nonZeroBins),3);

	# Compute the limiting mass at this redshift.
	my $limitingMass = 
	    +$coeffs->((0))
	    +$coeffs->((1))*$limitingMagnitude
	    +$coeffs->((2))*$limitingMagnitude**2;

	# Store the results.
	$redshiftTable     = $redshiftTable    ->append($redshift    );
	$limitingMassTable = $limitingMassTable->append($limitingMass);

    }
}

# Generate a fit to the data.
(my $fit, my $coeffs) = fitpoly1d($limitingMassTable,$redshiftTable,5);
my $fitMass     = sequence(1000)*5.0/999.0+8.0;
my $fitRedshift = pdl zeroes(nelem($fitMass));
my $closing     = "";
for(my $i=0;$i<nelem($coeffs);++$i) {
    $fitRedshift += $coeffs->(($i))*($fitMass**$i);
    print $coeffs->(($i))."d0";
    if ( $i == nelem($coeffs)-1 ) {
	print $closing."\n";
    } else {
	print "+logarithmicMass*(";
	$closing .= ")";
    }
}

# Create a plot.
my $plot;
my $gnuPlot;
my $plotFile = $workDirectory."massLuminosityRelation.pdf";
(my $plotFileEPS = $plotFile) =~ s/\.pdf$/.eps/;
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal epslatex color colortext lw 2 solid 7\n";
print $gnuPlot "set output '".$plotFileEPS."'\n";
print $gnuPlot "set title 'Redshift vs. Limiting Stellar Mass for Li \\& White (2009) Sample'\n";
print $gnuPlot "set xlabel 'Limiting stellar mass; \$M_\\star [M_\\odot]\$'\n";
print $gnuPlot "set ylabel 'Redshift; \$z\$'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.275,0.76\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale x\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [1.0e8:1.0e13]\n";
print $gnuPlot "set yrange [0.0:0.5]\n";
print $gnuPlot "set pointsize 2.0\n";
&PrettyPlots::Prepare_Dataset(
    \$plot,
    10.0**$fitMass,
    $fitRedshift,
    style      => "line",
    weight     => [5,3],
    color      => $PrettyPlots::colorPairs{'mediumSeaGreen'},
    title      => 'Fit'
    );
&PrettyPlots::Prepare_Dataset(
    \$plot,
    10.0**$limitingMassTable,
    $redshiftTable,
    style      => "point",
    weight     => [5,3],
    symbol     => [6,7],
    color      => $PrettyPlots::colorPairs{'redYellow'},
    title      => 'De Lucia (2006) SAM'
    );
&PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&LaTeX::GnuPlot2PDF($plotFileEPS);

exit;

File: constraints/dataAnalysis/stellarMassFunction_SDSS_z0.07/massLuminosityRelation.py
#!/usr/bin/env python3
# Determine the relation between stellar mass and SDSS r-band absolute magnitude using semi-analytic models from the Millennium
# Database. Specifically, the DeLucia2006a models are used - these correspond to De Lucia & Blaizot (2007;
# http://adsabs.harvard.edu/abs/2007MNRAS.375....2D) which use a Chabrier IMF, consistent with the IMF used by Li & White (2009).
# Andrew Benson (10-July-2012)

import os
import sys
import subprocess
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from astropy.cosmology import LambdaCDM
import astropy.units as u

execPath = os.environ.get('GALACTICUS_EXEC_PATH', '').rstrip('/') + '/'

# Define a working directory.
workDirectory = execPath + 'constraints/dataAnalysis/stellarMassFunction_SDSS_z0.07/massLuminosityWork/'
os.makedirs(workDirectory, exist_ok=True)

# Define little Hubble parameter for the Millennium Simulation.
hubble = 0.7

# Define the limiting apparent magnitude of the Li & White (2009) sample.
apparentMagnitudeLimit = 17.6

# Define Millennium Simulation cosmology.
cosmology = LambdaCDM(H0=70.0, Om0=0.3, Ode0=0.7)

# Define the database URL.
databaseURL = 'http://gavo.mpa-garching.mpg.de/Millennium/?action=doQuery&SQL='

# Get list of snapshot numbers and corresponding redshifts.
snapshotsFile = workDirectory + 'millenniumDB_snapshots.csv'
if not os.path.exists(snapshotsFile):
    sqlQuery = 'select snapnum, z from millimil..Snapshots'
    subprocess.run(['wget', databaseURL + sqlQuery, '-O', snapshotsFile], check=True)

snapshotNumbers = []
redshifts       = []
with open(snapshotsFile, 'r') as f:
    for line in f:
        if line.startswith('#') or 'snapnum' in line:
            continue
        parts = line.strip().split(',')
        if len(parts) >= 2:
            try:
                snapshotNumbers.append(int  (parts[0]))
                redshifts      .append(float(parts[1]))
            except ValueError:
                continue
snapshotNumbers = np.array(snapshotNumbers)
redshifts       = np.array(redshifts      )

# Initialize vectors to store results.
redshiftTable     = [0.0]
limitingMassTable = [6.0]

# Loop over snapshots in reverse order.
for iSnapshot in range(len(snapshotNumbers) - 1, -1, -1):
    redshift = redshifts[iSnapshot]

    # Only consider reasonably low redshifts.
    if 0.0 < redshift <= 0.5:

        # Compute the limiting absolute magnitude at this redshift.
        dL = cosmology.luminosity_distance(redshift).to(u.pc).value
        limitingMagnitude = apparentMagnitudeLimit - 5.0 * np.log10(dL / 10.0)

        # Define the SQL query.
        snapnum  = snapshotNumbers[iSnapshot]
        sqlQuery = (
            'select mass.stellarMass, light.r_sdss'
            ' from millimil..DeLucia2006a as mass, millimil..DeLucia2006a_sdss2mass as light'
            f' where mass.snapnum = {snapnum}'
            ' and mass.galaxyID = light.galaxyID'
        )

        # Download the data.
        csvFile = workDirectory + f'millenniumDB_{snapnum}.csv'
        if not os.path.exists(csvFile):
            subprocess.run(['wget', databaseURL + sqlQuery, '-O', csvFile], check=True)

        # Read the data.
        mass      = []
        magnitude = []
        with open(csvFile, 'r') as f:
            for line in f:
                if line.startswith('#') or 'stellarMass' in line:
                    continue
                parts = line.strip().split(',')
                if len(parts) >= 2:
                    try:
                        mass     .append(float(parts[0]))
                        magnitude.append(float(parts[1]))
                    except ValueError:
                        continue
        if not mass:
            continue
        mass      = np.array(mass     )
        magnitude = np.array(magnitude)

        # Convert mass to logarithmic scale and Solar units.
        logMass = np.log10(1.0e10 * mass / hubble)

        # Define mass bins.
        logMassMinimum = 8.0
        logMassMaximum = 13.0
        logMassCount   = 10
        logMassBins    = np.linspace(logMassMinimum, logMassMaximum, logMassCount)

        # Find the median magnitude as a function of mass.
        medianMag = np.zeros(logMassCount)
        for iBin in range(logMassCount):
            if iBin < logMassCount - 1:
                inBin = np.where((logMass >= logMassBins[iBin]) & (logMass < logMassBins[iBin + 1]))[0]
            else:
                inBin = np.where(logMass >= logMassBins[iBin])[0]
            if len(inBin) > 0:
                medianMag[iBin] = np.median(magnitude[inBin])

        # Fit a degree-3 polynomial to magnitude vs. mass (only for bins with negative magnitude).
        nonZero = np.where(medianMag < 0.0)[0]
        if len(nonZero) < 4:
            continue
        coeffsDescending = np.polyfit(medianMag[nonZero], logMassBins[nonZero], 3)
        coeffsAscending  = coeffsDescending[::-1]

        # Compute the limiting mass at this redshift.
        limitingMass = (
            coeffsAscending[0] +
            coeffsAscending[1] * limitingMagnitude +
            coeffsAscending[2] * limitingMagnitude**2
        )

        redshiftTable    .append(redshift    )
        limitingMassTable.append(limitingMass)

redshiftTable     = np.array(redshiftTable    )
limitingMassTable = np.array(limitingMassTable)

# Generate a fit to the data (degree 5).
coeffsDescending = np.polyfit(limitingMassTable, redshiftTable, 5)
coeffsAscending  = coeffsDescending[::-1]
fitMass = np.linspace(8.0, 13.0, 1000)
fitRedshift = np.zeros(len(fitMass))
closing = ''
for i, c in enumerate(coeffsAscending):
    fitRedshift += c * fitMass ** i
    print(f'{c:.10e}d0', end='')
    if i == len(coeffsAscending) - 1:
        print(closing)
    else:
        print('+logarithmicMass*(', end='')
        closing += ')'

# Create a plot.
fig, ax = plt.subplots(figsize=(8, 6))
ax.set_title(r'Redshift vs. Limiting Stellar Mass for Li \& White (2009) Sample')
ax.set_xlabel(r'Limiting stellar mass; $M_\star\;[M_\odot]$')
ax.set_ylabel(r'Redshift; $z$')
ax.set_xscale('log')
ax.set_xlim(1.0e8, 1.0e13)
ax.set_ylim(0.0, 0.5)
ax.plot(10.0 ** fitMass, fitRedshift,
        color='mediumseagreen', linewidth=2.5, label='Fit')
ax.plot(10.0 ** limitingMassTable, redshiftTable,
        'o', color='goldenrod', markersize=8, label='De Lucia (2006) SAM')
ax.legend(loc='lower left')
plt.tight_layout()
plotFile = workDirectory + 'massLuminosityRelation.pdf'
plt.savefig(plotFile)
plt.close()
